Cython/Compiler/ExprNodes.py

```diff
@@ -11364,20 +11364,43 @@ def compute_c_result_type(self, type1, type2):
                 self, type1, type2)
 
     def py_operation_function(self, code):
-        is_unicode_concat = False
-        if isinstance(self.operand1, FormattedValueNode) or isinstance(self.operand2, FormattedValueNode):
-            is_unicode_concat = True
+        type1, type2 = self.operand1.type, self.operand2.type
+        func = None
+        if type1 is unicode_type or type2 is unicode_type:
+            if type1 in (unicode_type, str_type) and type2 in (unicode_type, str_type):
+                is_unicode_concat = True
+            elif isinstance(self.operand1, FormattedValueNode) or isinstance(self.operand2, FormattedValueNode):
+                # Assume that even if we don't know the second type, it's going to be a string.
+                is_unicode_concat = True
+            else:
+                # Operation depends on the second type.
+                is_unicode_concat = False
+
+            if is_unicode_concat:
+                func = '__Pyx_PyUnicode_Concat'
+        elif type1 is bytes_type and type2 is bytes_type:
+            func = '__Pyx_PyBytes_Concat'
+
+        if func:
+            # any necessary utility code will be got by "NumberAdd" in generate_evaluation_code
+            if self.inplace or self.operand1.is_temp:
+                func = func.upper() + 'INPLACE'  # upper case to indicate unintuitive macro
+            if self.operand1.may_be_none() or self.operand2.may_be_none():
+                func += 'Safe'
+            return func
+
+        if self.operand1.is_temp:
+            # a specific optimization for string concatenation (where the type isn't known in advance)
+            return "__PYX_PYNUMBER_ADDTOTEMP"
```

Comment:
I'd rather not risk degrading the performance of the `+` operator, even by a tiny bit. It is used much more often for numbers than for strings, I'm sure.